tasks/f3/f3_task.go

```diff
@@ -0,0 +1,281 @@
+	for stillOwned() {
+		manifest, err := f.api.F3GetManifest(ctx)
+		switch {
+		case errors.Is(err, api.ErrF3Disabled):
+			log.Errorw("Cannot await F3 participation lease expiry as F3 is disabled.", "err", err)
+			return xerrors.Errorf("awaiting F3 participation lease expiry: %w", err)
+		case err != nil:
+			if backoff.Attempt() > ParticipationCheckProgressMaxAttempts {
+				log.Errorw("Too many failures while attempting to check F3  progress. Restarting participation.", "attempts", backoff.Attempt(), "err", err)
+				return nil
+			}
+			log.Errorw("Failed to check F3 progress while awaiting lease expiry. Retrying after backoff.", "attempts", backoff.Attempt(), "backoff", backoff.Duration(), "err", err)
+			time.Sleep(backoff.Duration())
+		case manifest == nil || manifest.NetworkName != lease.Network:
+			// If we got an unexpected manifest, or no manifest, go back to the
+			// beginning and try to get another ticket. Switching from having a manifest
+			// to having no manifest can theoretically happen if the lotus node reboots
+			// and has no static manifest.
+			return nil
+		}
+		switch progress, err := f.api.F3GetProgress(ctx); {
+		case errors.Is(err, api.ErrF3Disabled):
+			log.Errorw("Cannot await F3 participation lease expiry as F3 is disabled.", "err", err)
+			return xerrors.Errorf("awaiting F3 participation lease expiry: %w", err)
+		case err != nil:
+			if backoff.Attempt() > ParticipationCheckProgressMaxAttempts {
+				log.Errorw("Too many failures while attempting to check F3  progress. Restarting participation.", "attempts", backoff.Attempt(), "err", err)
+				return nil
+			}
+			log.Errorw("Failed to check F3 progress while awaiting lease expiry. Retrying after backoff.", "attempts", backoff.Attempt(), "backoff", backoff.Duration(), "err", err)
+			time.Sleep(backoff.Duration())
+		case progress.ID+renewLeaseWithin >= lease.ToInstance():
+			log.Infof("F3 progressed (%d) to within %d instances of lease expiry (%d). Renewing participation.", progress.ID, renewLeaseWithin, lease.ToInstance())
+			return nil
+		default:
+			remainingInstanceLease := lease.ToInstance() - progress.ID
+			waitTime := time.Duration(remainingInstanceLease-renewLeaseWithin) * manifest.CatchUpAlignment
+			if waitTime == 0 {
+				waitTime = 100 * time.Millisecond
+			}
+			log.Debugf("F3 participation lease is valid for further %d instances. Re-checking after %s.", remainingInstanceLease, waitTime)
+			time.Sleep(waitTime)
+		}
+	}
+	return ctx.Err()
+}
+
+func (f *F3Task) CanAccept(ids []harmonytask.TaskID, engine *harmonytask.TaskEngine) (*harmonytask.TaskID, error) {
+	id := ids[0]
+	return &id, nil
+}
+
+func (f *F3Task) TypeDetails() harmonytask.TaskTypeDetails {
+	return harmonytask.TaskTypeDetails{
+		Name: taskhelp.BackgroundTask("F3Participate"),
+		Cost: resources.Resources{
+			Cpu: 0,
+			Gpu: 0,
+			Ram: 10 << 20,
+		},
+		MaxFailures: 1,
+	}
+}
+
+func (f *F3Task) Adder(taskFunc harmonytask.AddTaskFunc) {
```

Comment:
Where is this added function called from? Did I miss the scheduling part somewhere?

Reply:
This is the normal adder func called by harmonytask on startup